tensorflow/compiler/mlir/hlo/lib/Dialect/mhlo/transforms/ral_legalize_to_llvm.cc

```diff
@@ -0,0 +1,556 @@
+  auto kernel_module = SymbolTable::lookupNearestSymbolFrom<gpu::GPUModuleOp>(
+      launch_op, launch_op.getKernelModuleName());
+  if (!kernel_module) {
+    launch_op.emitOpError() << "cannot find corresponding kernel module.";
+    return failure();
+  }
+
+  auto binary_attr =
+      kernel_module->getAttrOfType<StringAttr>(kGpuBinaryAttrName);
+  if (!binary_attr) {
+    kernel_module.emitOpError()
+        << "missing " << kGpuBinaryAttrName << " attribute";
+    return failure();
+  }
+
+  Operation* op = launch_op.getOperation();
+  Location loc = launch_op.getLoc();
+
+  // Create a global for the module blob.
+  StrT name_buffer(kernel_module.getName());
+  name_buffer.append("_blob");
+
+  Value module_blob = loadOrCreateGlobalString(rewriter, op, name_buffer.str(),
+                                               binary_attr.getValue());
+
+  // Make sure the trailing zero is included in the constant.
+  auto kernel_name = launch_op.getKernelName();
+  SmallString<128> kernel_name_buffer(kernel_name);
+  kernel_name_buffer.push_back('\0');
+
+  // Create a global for the kernel name.
+  SmallString<128> kernel_name_global_name_buffer;
+  auto kernel_name_global_name =
+      (kernel_module.getName() + "_" + kernel_name + "_kernel_name")
+          .toStringRef(kernel_name_global_name_buffer);
+  Value kernel_name_global = loadOrCreateGlobalString(
+      rewriter, op, kernel_name_global_name, kernel_name_buffer.str());
+
+  auto adaptor =
+      gpu::LaunchFuncOpAdaptor(operands, launch_op->getAttrDictionary());
+
+  // The Ral Context is the first argument of the surrounding LLVMFunc.
+  Value context_arg =
+      launch_op->getParentOfType<LLVM::LLVMFuncOp>().getArgument(0);
+  auto kernel_params = generateParamsArray(launch_op, operands, rewriter);
+
+  Type llvm_int32_type = IntegerType::get(rewriter.getContext(), 32);
+  Value zero = rewriter.create<LLVM::ConstantOp>(loc, llvm_int32_type,
+                                                 rewriter.getI32IntegerAttr(0));
+  // clang-format off
+  // TODO(disc): we use the default stream a.t.m. Implement a stream assignment algo in case necessary.
+  SmallVector<Value, 12> newOperands{
+      module_blob, /* gpu module string */
+      kernel_name_global, /* name of the kernel to launch */
+      adaptor.gridSizeX(), adaptor.gridSizeY(), adaptor.gridSizeZ(),
+      adaptor.blockSizeX(), adaptor.blockSizeY(), adaptor.blockSizeZ(),
+      zero, /* sharedMemBytes */
+      zero, /* gpu stream index */
+      kernel_params /* params for the kernel to launch */
+  };
+  // clang-format on
+
+  rewriter.replaceOpWithNewOp<disc_ral::DispatchOp>(
+      launch_op, llvm::None, context_arg, newOperands, kRalGpuLaunch, false,
+      "cpu");
+  return success();
+}
+
+class RalToLLVMPass : public RalToLLVMPassBase<RalToLLVMPass> {
+  void getDependentDialects(DialectRegistry& registry) const override {
+    registry.insert<LLVM::LLVMDialect>();
+  }
```

Comment:
You can put this information in the `.td` file and it won't be needed here.

Reply:
Thanks! It will introduce `llvm dialect` dependency into header file (e.g. `PassDetail.h`) if we put this dialect dependency in `td` file since some passes may shared one `.td` file.

Reply:
OK, looks fine to leave it there.